cpp/log/cms_verifier.cc

```diff
@@ -98,6 +168,49 @@ util::Status CmsVerifier::UnpackCmsDerBio(BIO* cms_bio_in, const Cert& cert,
 }
 
 
+util::Status CmsVerifier::UnpackCmsDerBio(BIO* cms_bio_in, BIO* cms_bio_out) {
+  CHECK_NOTNULL(cms_bio_in);
+  CHECK_NOTNULL(cms_bio_out);
+
+  CMS_ContentInfo* const cms_content_info = d2i_CMS_bio(cms_bio_in, nullptr);
+
+  if (!cms_content_info) {
+    LOG(ERROR) << "Could not parse CMS data";
+    LOG_OPENSSL_ERRORS(WARNING);
+    return Status(util::error::INVALID_ARGUMENT,
+                  "CMS data could not be parsed");
+  }
+
+  const ASN1_OBJECT* message_content_type(
+      CMS_get0_eContentType(cms_content_info));
+  int content_type_nid = OBJ_obj2nid(message_content_type);
+  // TODO: Enforce content type here. This is not yet defined in the RFC.
+  if (content_type_nid != NID_ctV2CmsPayloadContentType) {
+    LOG(WARNING) << "CMS message content has unexpected type: "
+                 << content_type_nid;
+  }
+
+  // Must set CMS_NOINTERN as the RFC says certs SHOULD be omitted from the
+  // message but the client might not have obeyed this. CMS_BINARY is required
+  // to avoid MIME-related translation. CMS_NO_SIGNER_CERT_VERIFY because we
+  // will do our own checks that the chain is valid and the message may not
+  // be signed directly by a trusted cert. CMS_NO_CONTENT_VERIFY because we
+  // can't apply the RFC mandated signature checks until we have the unpacked
+  // cert to examine. We don't check it's a signed data object CMS type as
+  // OpenSSL does this.
+  int verified = CMS_verify(cms_content_info, nullptr, nullptr,
+                            nullptr, cms_bio_out,
+                            CMS_NO_SIGNER_CERT_VERIFY | CMS_NOINTERN
+                            | CMS_BINARY | CMS_NO_CONTENT_VERIFY);
+
+  CMS_ContentInfo_free(cms_content_info);
+
+  return (verified == 1) ? util::Status::OK
+                         : util::Status(util::error::INVALID_ARGUMENT,
+                                        "CMS unpack failed");
```

Comment:
I didn't want to use "verify" here as it's just unpacking. The checks come later.